File: F#/Test/dequeue.fs
module FSharp.DequeueTest

open FsUnit
open FSharp.Dequeue
open NUnit.Framework

let [<Test>] size () =
    let dequeue = Dequeue()

    dequeue.size() |> should equal 0
    
    dequeue.addFront 1

    dequeue.size() |> should equal 1


let [<Test>] addFront () =
    let dequeue = Dequeue()

    dequeue.addFront 1
    dequeue.addFront 2
    dequeue.addFront 3

    dequeue.removeRear() |> should equal 1


let [<Test>] addRear () =
    let dequeue = Dequeue()

    dequeue.addRear 1
    dequeue.addRear 2
    dequeue.addRear 3

    dequeue.removeRear() |> should equal 3


let [<Test>] removeFront () =
    let dequeue = Dequeue()

    dequeue.addRear 1
    dequeue.addRear 2
    dequeue.addRear 3

    dequeue.removeFront() |> should equal 1
    dequeue.removeFront() |> should equal 2
    dequeue.removeFront() |> should equal 3
    dequeue.size() |> should equal 0

File: C++/tests/dequeue.cpp
//
// Created by Daniel on 06.04.2015.
//

#include "gtest/gtest.h"
#include "../dequeue.h"

TEST(Dequeue, size) {
    auto testee = dequeue<int>();
    ASSERT_EQ(0u, testee.size());
    testee.addFront(1);
    ASSERT_EQ(1u, testee.size());
}


TEST(Dequeue, addFront) {
    auto testee = dequeue<int>();
    testee.addFront(1);
    testee.addFront(2);
    testee.addFront(3);
    ASSERT_EQ(1, testee.removeRear());
}


TEST(Dequeue, addRear) {
    auto testee = dequeue<int>();
    testee.addRear(1);
    testee.addRear(2);
    testee.addRear(3);

    ASSERT_EQ(3, testee.removeRear());
}


TEST(Dequeue, removeFront) {
    auto testee = dequeue<int>();
    testee.addRear(1);
    testee.addRear(2);
    testee.addRear(3);

    ASSERT_EQ(1, testee.removeFront());
    ASSERT_EQ(2, testee.removeFront());
    ASSERT_EQ(3, testee.removeFront());
    ASSERT_EQ(0u, testee.size());
}